src/main/scala/ch/zzeekk/spark/temporalquery/IntervalDef.scala

```diff
@@ -0,0 +1,196 @@
+
+  /**
+   * Define min value of the interval axis
+   */
+  def minValue: T
+  /**
+   * Define max value of the interval axis
+   */
+  def maxValue: T
+  /**
+   * Expression to check if a value is included in a given interval
+   */
+  def isInIntervalExpr(valueCol: Column, fromCol: Column, toCol: Column): Column
+  /**
+   * Expression to check if an interval is valid, e.g. start is before end.
+   */
+  def isValidIntervalExpr(fromCol: Column, toCol: Column): Column
+  /**
+   * Expression to join two intervals
+   */
+  def intervalJoinExpr(fromCol1: Column, toCol1: Column, fromCol2: Column, toCol2: Column): Column
+  /**
+   * Method to floor a scala value of type T
+   */
+  def floor(value: T): T
+  /**
+   * Method to ceil a scala value of type T
+   */
+  def ceil(value: T): T
+  /**
+   * Method get the predecessor for a scala value of type T for this interval axis definition
+   */
+  def predecessor(value: T): T
+  /**
+   * Method get the successor for a scala value of type T for this interval axis definition
+   */
+  def successor(value: T): T
+
+  /**
+   * Expression to floor a column
+   */
+  def getFloorExpr(valueCol: Column): Column = {
+    val udfTransform = udf((v: Any) => Option(v).map(x => floor(x.asInstanceOf[T])))
+    udfTransform(valueCol)
+  }
+  /**
+   * Expression to ceil a column
+   */
+  def getCeilExpr(valueCol: Column): Column = {
+    val udfTransform = udf((v: Any) => Option(v).map(x => ceil(x.asInstanceOf[T])))
+    udfTransform(valueCol)
+  }
+  /**
+   * Expression get predecessor for a column
+   */
+  def getPredecessorExpr(valueCol: Column): Column = {
+    val udfTransform = udf((v: Any) => Option(v).map(x => predecessor(x.asInstanceOf[T])))
+    udfTransform(valueCol)
+  }
+  /**
+   * Expression get successor for a column
+   */
+  def getSuccessorExpr(valueCol: Column): Column = {
+    val udfTransform = udf((v: Any) => Option(v).map(x => successor(x.asInstanceOf[T])))
+    udfTransform(valueCol)
+  }
+  /**
+   * make sure value is between min- and maxValue
+   */
+  @inline def fitToBoundaries(value: T): T = least(greatest(value, minValue), maxValue)
+  def getFitToBoundariesExpr(valueCol: Column): Column = {
+    functions.when(valueCol.isNotNull, functions.least(functions.greatest(valueCol, functions.lit(minValue)), functions.lit(maxValue)))
+  }
+
+  // Helpers
+  @inline private def least(values: T*): T = values.min
+  @inline private def greatest(values: T*): T = values.max
+}
+
+/**
+ * A closed interval is an interval which includes its lower and upper bound.
+ * Use this for discret interval axis.
+ * @param minValue minimum value of the interval axis
+ * @param maxValue maximum value of the interval axis
+ * @tparam T: scala type for discrete interval axis, e.g. Timestamp, Integer, ...
+ */
+case class ClosedInterval[T: Ordering: TypeTag](override val minValue: T, override val maxValue: T, discreteAxisDef: DiscreteAxisDef[T]) extends IntervalDef[T] {
```

Comment:
```suggestion
case class ClosedInterval[T: Ordering: Integral: TypeTag](override val minValue: T, override val maxValue: T, discreteAxisDef: DiscreteAxisDef[T]) extends IntervalDef[T] {
```
You need to exclude Fractional, i.e. types with a dense ordering because the implementation of `predecessor` et al. assumes a discrete type T.

Reply:
Well, they are excluded by the DiscreteAxis implementations available:
- case class DiscreteTimeAxis(timeUnit: ChronoUnit) extends DiscreteAxisDef[Timestamp]
- case class DiscreteNumericAxis[T: Integral](step: T)(implicit f: Integral[T]) extends DiscreteAxisDef[T]

Unfortunately Timestamp is no Integral, so we cant constrain it already on ClosedInterval...

Reply:
It might be possible to provide an Integral implementation for Timestamp by converting it to a Long, but it's not obvious for me to implement.
The current implementation is constraint by the DiscreteAxis implementations available, i suggest to leave it as-is.